src/blindrsa/blindrsa.ts

```diff
@@ -0,0 +1,157 @@
+        r_inv = r.inverseMod(new sjcl.bn(n));
+    } catch (e) {
+        throw new Error('invalid blind');
+    }
+    // 7. x = RSAVP1(pkS, r)
+    const x = rsavp1({ n, e }, r);
+
+    // 8. z = m * x mod n
+    const z = m.mulmod(x, n);
+
+    // 9. blinded_msg = int_to_bytes(z, kLen)
+    const blindedMsg = i2osp(z, kLen);
+
+    // 10. inv = int_to_bytes(r_inv, kLen)
+    const blindInv = i2osp(r_inv, kLen);
+
+    // 11. output blinded_msg, inv
+    return { blindedMsg, blindInv };
+}
+
+export async function finalize(
+    publicKey: CryptoKey,
+    msg: Uint8Array,
+    blindInv: Uint8Array,
+    blindSig: Uint8Array,
+    saltLength = 0,
+): Promise<Uint8Array> {
+    if (publicKey.type !== 'public' || publicKey.algorithm.name !== 'RSA-PSS') {
+        throw new Error('key is not RSA-PSS');
+    }
+    if (!publicKey.extractable) {
+        throw new Error('key is not extractable');
+    }
+    const { modulusLength } = publicKey.algorithm as RsaHashedKeyGenParams;
+    const kLen = Math.ceil(modulusLength / 8);
+
+    // 1. If len(blind_sig) != kLen, raise "unexpected input size" and stop
+    // 2. If len(inv) != kLen, raise "unexpected input size" and stop
+    if (blindSig.length != kLen || blindInv.length != kLen) {
+        throw new Error('unexpected input size');
+    }
+
+    // 3. z = bytes_to_int(blind_sig)
+    const z = os2ip(blindSig);
+
+    // 4. r_inv = bytes_to_int(inv)
+    const r_inv = os2ip(blindInv);
+
+    // 5. s = z * r_inv mod n
+    const jwkKey = await crypto.subtle.exportKey('jwk', publicKey);
+    if (!jwkKey.n) {
+        throw new Error('key has invalid parameters');
+    }
+    const n = new sjcl.bn(Buffer.from(jwkKey.n, 'base64url').toString('hex'));
+    const s = z.mulmod(r_inv, n);
+
+    // 6. sig = int_to_bytes(s, kLen)
+    const sig = i2osp(s, kLen);
+
+    // 7. result = RSASSA-PSS-VERIFY(pkS, msg, sig)
+    // 8. If result = "valid signature", output sig, else
+    //    raise "invalid signature" and stop
+    const algorithm = { name: 'RSA-PSS', saltLength };
+    if (!(await crypto.subtle.verify(algorithm, publicKey, sig, msg))) {
+        throw new Error('invalid signature');
+    }
+
+    return sig;
+}
+
+export async function blindSign(privateKey: CryptoKey, blindMsg: Uint8Array): Promise<Uint8Array> {
+    if (privateKey.type !== 'private' || privateKey.algorithm.name !== 'RSA-PSS') {
+        throw new Error('key is not RSA-PSS');
+    }
+    if (!privateKey.extractable) {
+        throw new Error('key is not extractable');
+    }
+    const { modulusLength } = privateKey.algorithm as RsaHashedKeyGenParams;
+    const kLen = Math.ceil(modulusLength / 8);
+
+    const jwkKey = await crypto.subtle.exportKey('jwk', privateKey);
+    if (!jwkKey.n || !jwkKey.d) {
+        throw new Error('key is not a private key');
+    }
+    const n = new sjcl.bn(Buffer.from(jwkKey.n, 'base64url').toString('hex'));
+    const d = new sjcl.bn(Buffer.from(jwkKey.d, 'base64url').toString('hex'));
```

Comment:
Minor nit: at https://github.com/privacypass/challenge-bypass-extension/pull/328/files#diff-80aac2cdd6b81297ebba46b540a67548f0013f4c2b574e281fc716bad65b900aR32-R37 the calculation of these intermediate values is delayed until just before their first use. If we wanted to do the same here, we could move lines https://github.com/privacypass/challenge-bypass-extension/pull/328/files#diff-80aac2cdd6b81297ebba46b540a67548f0013f4c2b574e281fc716bad65b900aR130-R135 to after step 2 below for consistency.

Reply:
done